### PR TITLE
[Transaction Async get] Part 2: Refactored TransactionService and futures helper.

### PR DESCRIPTION
**Pull request by OStevan**
**Goals (and why)**:
Refactor `TransactionService` for easier addition of async paths.

**Implementation Description (bullets)**:
- same trick used as before (wrap in `Futures.immediateFuture`)

**Testing (What was existing testing like?  What have you done to improve it?)**:
- existing tests should cover this

**Concerns (what feedback would you like?)**:
- is there a way to make loading/getting interfaces sharable

**Where should we start reviewing?**:
- TransactionService

**Priority (whenever / two weeks / yesterday)**:
- this week

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/PreStartHandlingTransactionService.java [hunk 1]
```diff
@@ -44,33 +49,27 @@
  */
 public class PreStartHandlingTransactionService implements TransactionService {
     private final TransactionService delegate;
+    private final TimestampLoader immediateTimestampLoader;
 
-    public PreStartHandlingTransactionService(TransactionService delegate) {
+    public static TransactionService create(TransactionService delegate) {
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/PreStartHandlingTransactionService.java]: Inlined and removed.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/service/SplitKeyDelegatingTransactionServiceTest.java [hunk 2]
```diff
@@ -82,13 +84,15 @@ public void getThrowsIfFunctionReturnsUnmappedValue() {
...
         RuntimeException ex = new IllegalStateException("bad");
-        TransactionService unusableService = new SplitKeyDelegatingTransactionService<>(
+        TransactionService unusableService = SplitKeyDelegatingTransactionService.create(
                 num -> {
                     throw ex;
                 },
                 transactionServiceMap);
 
-        assertThatThrownBy(() -> unusableService.get(5L)).isEqualTo(ex);
+        assertThatThrownBy(() -> unusableService.get(5L))
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/service/SplitKeyDelegatingTransactionServiceTest.java]: Something with `runWithUnchecked` made the change necessary. After refactoring is no longer a problem.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/futures/AtlasFutures.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import com.palantir.common.base.Throwables;
+
+public final class AtlasFutures {
+    private AtlasFutures() {
+
+    }
+
+    public static <R> R runWithException(FutureCallable<R> futureCallable) {
+        try {
+            return futureCallable.call().get();
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/futures/AtlasFutures.java]: Implemented a `getUnchecked` for our specific case.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java [hunk 4]
```diff
@@ -114,4 +100,45 @@ public TimestampEncodingStrategy getEncodingStrategy() {
...
+
+    private ListenableFuture<Map<Long, Long>> getInternal(Iterable<Long> startTimestamps, CellLoader cellLoader) {
+        Map<Cell, Long> startTsMap = Maps.newHashMap();
+        for (Long startTimestamp : startTimestamps) {
+            Cell cell = getTransactionCell(startTimestamp);
+            startTsMap.put(cell, MAX_TIMESTAMP);
+        }
+
+        return Futures.transform(cellLoader.get(startTsMap),
+                rawResults -> {
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java]: Extracted.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java [hunk 5]
```diff
@@ -114,4 +100,45 @@ public TimestampEncodingStrategy getEncodingStrategy() {
...
+                                returnMap.get(cell).getContents());
+                    } else {
+                        return null;
+                    }
+                },
+                MoreExecutors.directExecutor());
+    }
+
+    private ListenableFuture<Map<Long, Long>> getInternal(Iterable<Long> startTimestamps, CellLoader cellLoader) {
+        Map<Cell, Long> startTsMap = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java]: Extracted.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SplitKeyDelegatingTransactionService.java [hunk 6]
```diff
@@ -82,23 +124,16 @@ public Long get(long startTimestamp) {
...
-    }
-
-    @Override
-    public void putUnlessExists(long startTimestamp, long commitTimestamp) throws KeyAlreadyExistsException {
-        TransactionService service = getServiceForTimestamp(startTimestamp).orElseThrow(
-                () -> new UnsupportedOperationException("putUnlessExists shouldn't be used with null services"));
-        service.putUnlessExists(startTimestamp, commitTimestamp);
-    }
+        List<ListenableFuture<Map<Long, Long>>> futures = KeyedStream.stream(queryMap.asMap())
+                .entries()
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SplitKeyDelegatingTransactionService.java]: Refactored.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java [hunk 7]
```diff
@@ -114,4 +100,45 @@ public TimestampEncodingStrategy getEncodingStrategy() {
...
+                                .decodeValueAsCommitTimestamp(startTs, e.getValue().getContents());
+                        result.put(startTs, commitTs);
+                    }
+
+                    return result;
+                },
+                MoreExecutors.directExecutor());
+    }
+
+    private interface CellLoader {
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SimpleTransactionService.java]: Renamed and should be resolved.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/PreStartHandlingTransactionService.java [hunk 8]
```diff
@@ -44,33 +49,27 @@
...
+        return new PreStartHandlingTransactionService(delegate, TransactionServices.immediateTimestampLoader());
+    }
+
+    PreStartHandlingTransactionService(TransactionService delegate,
+            TimestampLoader immediateTimestampLoader) {
         this.delegate = delegate;
+        this.immediateTimestampLoader = immediateTimestampLoader;
```
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/PreStartHandlingTransactionService.java]: If understood correctly refactored.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SplitKeyDelegatingTransactionService.java [hunk 9]
```diff
@@ -46,25 +52,51 @@
...
     private final Map<T, TransactionService> keyedServices;
+    private final Map<T, AsyncTransactionService> keyedAsyncServices;
 
-    public SplitKeyDelegatingTransactionService(
+    SplitKeyDelegatingTransactionService(
             Function<Long, T> timestampToServiceKey,
             Map<T, TransactionService> keyedServices) {
         this.timestampToServiceKey = timestampToServiceKey;
         this.keyedServices = keyedServices;
+        this.keyedAsyncServices = KeyedStream.stream(keyedServices)
```
Comment (felixdesouza) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/service/SplitKeyDelegatingTransactionService.java]: these are synchronous as opposed to async

**Pull request merged**